feat: implement v2 Transaction Types

sleepdefic1t opened the pull request:
This PR Implements v2 Transaction Types.

It also does the following:
- adds documentation and fixtures detailing the deserialization process of types 0-10 Transactions.
- adds examples for unimplemented operation types (MultiPayment, Message Signing, etc).
- bumps version to 1.0.0.
- updates the changelog.

## What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Does this PR release a new version?

- [x] Yes
- [ ] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `develop` branch, _not_ the `master` branch
- [x] All tests are passing
- [x] New/updated tests are included

sleepdefic1t committed:
feat: implement v2 Transaction Types

sleepdefic1t committed:
Update README.md

sleepdefic1t commented:
Full reimplementation of Legacy Transactions currently at [/develop](https://github.com/ArkEcosystem/ledger/tree/develop) via #25

vasild commented on docs/fixtures/type_0.md:
````diff
@@ -0,0 +1,51 @@
+ff0217010000000000010000000000000003a02b9d5fdd1307c2ee4652ba54d492d1fd11a7d1bb3f3a44c4a05e79f19de933809698000000000000a08601000000000000000000171dfc69b54c7fe901e91d5a9ab78388645e2427ea4f01bd21828a633a3c821b9984fe642deab87237b99e62a543ca6948ff1d6d32f2475ada1f933da0591c40603693614afa69fcb4caa2b4be018788de9f10c42a
+```
+
+## Deserializing the Hex
````
:1st_place_medal:

vasild commented on docs/fixtures/type_7.md:
```diff
@@ -0,0 +1,46 @@
+| **Type:**                 | **[7]**   | **2**             | `0x0700`              |
+| **Nonce:**                | **[9]**   | **8**             | `0x0200000000000000`  |
+| **SenderPublicKey:**      | **[17]**  | **33**            | `0x037a12518205254e6ebf25290d9786fd9821c43bb7319c9fc2499c8d472809dfaf`    |
+| **Fee:**                  | **[50]**  | **8**             | `0x00f9029500000000`  |
```
no "Amount"?

sleepdefic1t replied:
It's not serialized, or at least it's not in the payload.
If that's not an error, I can add a line stating as such 👍 

`| **Amount:**               | **[..]**  | **0**             |  Not Serialized       |`

spkjp replied:
Yeah, we do not serialize the "total amount" of a multipayment.

vasild commented on docs/unimplemented/type_2/type_2.c:
```diff
@@ -0,0 +1,157 @@
+                                             const uint8_t *buffer,
+                                             const uint32_t length) {
+/////////
+    case TRANSACTION_TYPE_DELEGATE_REGISTRATION:
```
`case` without `switch`, I guess this and below is some draft that wouldn't compile.

sleepdefic1t replied:
Anything in the "unimplemented" folder just contains snippets and pieces of code where the methods could be implemented. It should not be compiled.

This specific part refers to a section in `/operations/transactions/deserializer.h`.

Basically, I added mostly everything, then removed the parts that wouldn't work on Ledger for either being too big or irrelevant.  
(e.g. no sense in registering a delegate on Ledger as you can't forge with a ledger).

vasild commented on docs/unimplemented/type_6/type_6.c:
```diff
@@ -0,0 +1,175 @@
+
+    for (uint8_t i = 0U; i < payments->n_payments; ++i) {
+        payments->amounts[i] = U8BE(&buffer[sizeof(uint16_t) +
+                                           (i * sizeof(uint64_t))],
```
You do not need the extra brackets: `A + (B * C)` is the same as `A + B * C`. The priority of `*` over `+` is well known to everybody. Redundant (unnecessary) brackets could be used for some subtle expressions that involve shift, bitwise operations or `?:`.

sleepdefic1t replied:
Kinda same as above, anything in the unimplemented folder isn't finished or working code, just snippets and examples of how and where one could implement the feat.

The brackets were left over from when I was working out the logic (which still may be incorrect), I'll remove them nonetheless and added notes at the top of every unimplemented feature.

vasild commented on docs/unimplemented/type_6/type_6.c:
```diff
@@ -0,0 +1,175 @@
+                                    sizeof(uint64_t));
+
+        os_memmove(&payments->addresses[i * ADDRESS_HASH_LENGTH],
+                   &buffer[(sizeof(uint16_t) + (i * (sizeof(uint64_t) + ADDRESS_HASH_LENGTH)))],
```
Similar to above, no need to have these brackets as they only clutter the code: `buffer[(...)]` is the same as `buffer[...]`

But more importantly, shouldn't this be
```
sizeof(uint16_t) + payments->n_payments * sizeof(uint64_t) + i * ADDRESS_HASH_LENGTH
```
instead of
```
sizeof(uint16_t) + (i * (sizeof(uint64_t) + ADDRESS_HASH_LENGTH))
```
?

sleepdefic1t replied:
😂 
Yes, yes is should.

I'll fix this snippet example too 👍

vasild commented on docs/unimplemented/type_6/type_6.c:
```diff
@@ -0,0 +1,175 @@
+////////////////////////////////////////////////////////////////////////////////
+
+const char * const ui_menu_multi_payment[][2] = {
+    { "Operation",      (const char *const)amountBuffer },
```
Inconsistency wrt spaces around `*` on line 76 vs lines 77-80.

sleepdefic1t replied:
I caught this in the parts that were implemented.
Even though it's only for reference, it's best it be as accurate as reasonably possible. 

I'll fix this in the updates 👍

vasild commented on src/operations/transactions/assets/type_0.c:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    transfer->amount = U8LE(buffer, 0U);
+    os_memmove(transfer->recipient,
```
What about `transfer->expiration`? It is in the comments above, but not set here.

sleepdefic1t replied:
Expiration isn't currently displayed on Ledger,
and there's no reason to deserialize and store what isn't being displayed.

When this was implemented,
it also had VF which would mean 6 screens on the Ledger, which I haven't tested yet.

VF is also currently unimplemented, and it may be easier to do 6-screens now that things have been whittled down though.

I'll add this back in for now 👍

vasild commented on src/operations/transactions/assets/type_0.c:
```diff
@@ -0,0 +1,68 @@
+StreamStatus deserializeTransfer(Transfer *transfer,
+                                 const uint8_t *buffer,
+                                 const uint32_t length) {
+    if (length != PUBLICKEY_COMPRESSED_LENGTH) {
```
Isn't `PUBLICKEY_COMPRESSED_LENGTH` the wrong constant to name "length of amount + length of expiration + length of recipient"?

sleepdefic1t replied:
It's definitely the wrong constant name and looks confusing.

The length is correct though.

Per AIP-11 doc, Transfer asset is 33 Bytes.
8 + 4 + 21.

I'll drop using the constant 👍

vasild replied:
Wow, what a coincidence! :)

vasild commented on src/operations/transactions/assets/type_5.c:
```diff
@@ -0,0 +1,69 @@
+
+    // Let's make sure the length isn't > 255,
+    // and that the lengths match.
+    if (!(length & 0xFF) || length != ipfs->length) {
```
`!(length & 0xFF)` is quite an obfuscated way to check whether `length > 255` and it will also evaluate to true if `length == 0`. Better use `length > 255` instead.

sleepdefic1t replied:
Will do, nice catch!
I was assuming 0 to be an invalid length, but the AIP11 doc says its valid.

Thank you, @vasild

vasild commented on src/operations/transactions/assets/types.h:
```diff
@@ -0,0 +1,84 @@
+                        ///////// Delegate Res.     // Type 7
+                        HTLC_LOCK = 9U,             // Type 8
+                        HTLC_CLAIM,                 // Type 9
+                        HTLC_REFUND                 // Type 10
```
Better assign value to every constant, to make it more readable. There is contradiction between the value `9U` and the comment `Type 8` on line 60.

sleepdefic1t replied:
Yup, that would definitely be incorrect.
I'll assign a value to every type 👍 
... with the correct numbers this time 😂

sleepdefic1t replied:
There's actually no reason to even have this.
Went to update and realized it ended up being unused.
Removed in updates.

vasild commented on src/operations/transactions/display.c:
```diff
@@ -0,0 +1,231 @@
+////////////////////////////////////////////////////////////////////////////////
+
+static void setDisplayTransfer(const Transaction *transaction) {
+    os_memmove((char *)displayCtx.operation, "Transfer\0", 9U);
```
The `\0` is unnecessary - `"Foo"` is the same as `"Foo\0"`.

sleepdefic1t replied:
Ahh, okay. that makes sense.
I was following the pattern of what was there before and didn't notice it's totally redundant.

I'll fix them all with these next updates 👍

ghost commented:
Your pull request needs some changes. Please wait for a comment from one of our developers for more information.

<!-- ACTION:REQUEST_CHANGES -->

sleepdefic1t committed:
fixes and updates

- adds more documentation to all unimplemented examples
- updates all unimplemented examples.
- fixes a few formatting issue.
- adds Expiration deserialization and Display to Transfer/Type 0.
- corrects IPFS DAG length checks.
- improves Tx Type objects.
  - no need for `TransactionTypes.
  - No need for hex in Tx Type constants.
  - improves readability.
- removes redundant '\0' on displayCtx title and vars.
- fixes a bug in adjustDecimals to allow printing number values where no Token name or decimal is needed.

sleepdefic1t committed:
remove more redundant null terminators

sleepdefic1t commented:
@vasild 
Updates done 👍

vasild commented:
lgtm

sleepdefic1t committed:
chore: update HTLC-related code and docs

- updates code, docs, and fixtures for HTLC-types.
- adds expirationType and expirationValue to HTLC Lock.
- adds display of HTLC Lock expiration.
- adds another (5th) UX Step.
- drops unused U8BE.
- minor formatting improvements.

sleepdefic1t committed:
chore: final touchups

- mostly formatting, naming, and unused header removal.
